### ISSUE TITLE
explicit_iter_loop auto-fix produces non-compiling code for returned closure (needs move)

### ISSUE THREAD

**Issue by abogaichuk** [rust-lang/rust-clippy#16207]
### Summary

The `clippy::explicit_iter_loop` lint’s auto-fix (via `cargo clippy --fix`) can produce **non-compiling code** when the loop is inside a closure that is returned from a function.

In my case, Clippy changes `for (xy, part) in grid.iter()` to `for (xy, part) in grid` inside a returned closure, which makes the closure borrow `grid` instead of capturing it by value. The closure escapes the function, so this triggers `E0373` (“closure may outlive the current function”).

The fix should either

* add `move` to the closure, or
* avoid applying `explicit_iter_loop` when it would cause this lifetime issue.

### Reproducer

Code:

```rust
pub fn closest_in_room_range(
    grid: &HashMap<RoomXY, RoomPart>,
) -> impl FnMut(RoomName, CostMatrix) -> SingleRoomCostResult + use<'_> {
    |_: RoomName, mut matrix: CostMatrix| -> SingleRoomCostResult {
        for (xy, part) in grid.iter() {
            ...
        }
        SingleRoomCostResult::CostMatrix(matrix)
    }
}
```

Current output:
```rust
pub fn closest_in_room_range(
    grid: &HashMap<RoomXY, RoomPart>,
) -> impl FnMut(RoomName, CostMatrix) -> SingleRoomCostResult + use<'_> {
    |_: RoomName, mut matrix: CostMatrix| -> SingleRoomCostResult {
        for (xy, part) in grid {
            ...
        }
        SingleRoomCostResult::CostMatrix(matrix)
    }
}
```

Desired output:
```rust
pub fn closest_in_room_range(
    grid: &HashMap<RoomXY, RoomPart>,
) -> impl FnMut(RoomName, CostMatrix) -> SingleRoomCostResult + use<'_> {
    move |_: RoomName, mut matrix: CostMatrix| -> SingleRoomCostResult {
        for (xy, part) in grid {
            ...
        }
        SingleRoomCostResult::CostMatrix(matrix)
    }
}
```

### Version

```text
rustc 1.91.1 (ed61e7d7e 2025-11-07)
binary: rustc
commit-hash: ed61e7d7e242494fb7057f2657300d9e77bb4fcb
commit-date: 2025-11-07
host: x86_64-unknown-linux-gnu
release: 1.91.1
LLVM version: 21.1.2
```
